README.md
# Samply Common LDM Client Centraxx

Samply Common LDM Client Centraxx is a library for the communication with **L**ocal **D**ata**m**anagement systems.
It extends the abstract Samply Common LDM Client.

## Features

Provides basic operations to communicate with Centraxx.

## Build

In order to build this project, you need to configure maven properly and use the maven profile that
fits to your project.

``` 
mvn clean package
```

## Configuration

Samply Common LDM Client Centraxx is configured during runtime by providing the necessary paramaters in the constructor.

For usage examples, see the test classes.

## Maven artifact

To use the module, include the following artifact in your project.

``` 
<dependency>
    <groupId>de.samply</groupId>
    <artifactId>common-ldmclient.centraxx</artifactId>
    <version>6.2.1</version>
</dependency>
``` 

 ## License
        
Copyright 2020 - 2021 The Samply Community
       
Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
       
http://www.apache.org/licenses/LICENSE-2.0
       
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).
## [6.2.1 - 2022-08-10]
- common ldm client version update

## [6.2.0 - 2021-12-30]
- Jakarta

## [6.1.3 - 2021-06-16]
- Remove PercentageLogger

## [6.1.2 - 2021-01-25]
- Google Guava Preconditions
- Percentage Logger start

## [6.1.1 - 2021-01-25]
- Guava Preconditions

## [6.1.0 - 2021-01-21]
- Percentage Logger

## [6.0.0 - 2020-11-09]
- Github release
### Changed
- Samply parent 11.1.0
## Added
- Github Actions
- Google Code Style

## [5.4.0 - 2020-04-24]
### Changed
- Update ldm-client (5.4.0)

## [5.3.0 - 2019-09-02]
### Changed
- Update dependencies

## [5.2.0 - 2019-08-26]
### Changed
- Update dependencies

## [5.1.0 - 2019-07-04]
### Changed
- Introduce LdmQueryResult as return value in LdmClient-API (replacing Object as return value)

## [5.0.0 - 2019-06-24]
### Changed
- Moved functionality to superclass LdmClient
- Cleaner code

## [4.0.1 - 2018-04-26]
### Changed
- samply parent 10.1 (Java 8)

## [4.0.0 - 2018-03-13]
### Changed
- samply parent 10.0 (Java 8)

## [3.0.1] - 2017-12-15
### Fixed
- Wrong class was used to check if it is a stats file or error file

## [3.0.0] - 2017-12-15
### Changed
- Adapt to v3(.0.1) of the interface ldmclient

## [2.0.0] - UNRELEASED
### Changed
- Adapt to v2 of the interface ldmclient

## [1.1.2] - 2017-10-09
### Changed
- Use QueryConverter from share.dto

### Fixed
- Check if cache is null before trying to invalitate it

## [1.1.1] - 2017-09-13
### Added
- Cache QueryResults (optional)

## [1.1.0] - Unreleased
- Skipped due to error in parent class

## [1.0.1] - 2017-04-27
### Added
- Get User Agent Info (ldm name + version info)

### Changed
- Increased timeouts when reading stats/error or results

## [1.0.0] - skipped
### Added
- First implementation


src/test/java/de/samply/common/ldmclient/centraxx/LdmClientCentraxxTest.java


package de.samply.common.ldmclient.centraxx;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assume.assumeFalse;

import com.google.common.base.Stopwatch;
import de.samply.common.ldmclient.AbstractLdmClient;
import de.samply.common.ldmclient.LdmClientUtil;
import de.samply.common.ldmclient.model.LdmQueryResult;
import de.samply.share.model.ccp.QueryResult;
import de.samply.share.model.common.View;
import java.io.IOException;
import java.io.InputStream;
import java.io.StringReader;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.TimeUnit;
import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Unmarshaller;
import org.apache.commons.io.IOUtils;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.xml.sax.InputSource;

/**
 * Test Class for the LDM Connector for Centraxx Backends
 */
public class LdmClientCentraxxTest {

  private static final Logger logger = LoggerFactory.getLogger(LdmClientCentraxxTest.class);
  private static final boolean CACHING = true;

  private static final String centraxxBaseUrl = "http://localhost:9090/centraxx/";

  // In order to avoid unnecessary creation of too many requests...use a pre-existing result id from centraxx. Check the queryresults folder on the centraxx machine
  private static final String predefinedResultId = "99bdb78a-4cd4-418b-b4de-95dcc331e83f";
  private static final String predefinedResultLocation =
      centraxxBaseUrl + AbstractLdmClient.REST_PATH_REQUESTS + "/" + predefinedResultId;

  // Same for errors
  private static final String predefinedErrorResultId = "0ebf8ebb-0132-46f6-aaf6-f682682279dd";
  private static final String predefinedErrorResultLocation =
      centraxxBaseUrl + AbstractLdmClient.REST_PATH_REQUESTS + "/" + predefinedErrorResultId;

  private static final String VIEW_FILENAME_OK = "exampleView.xml";

  private LdmClientCentraxx ldmClientCentraxx;
  private CloseableHttpClient httpClient;
  /*

  @Before
  public void setUp() throws Exception {
    this.httpClient = HttpClients.createDefault();
    this.ldmClientCentraxx = new LdmClientCentraxx(httpClient, centraxxBaseUrl, CACHING);
  }

  @After
  public void tearDown() throws Exception {
    httpClient.close();
    ldmClientCentraxx.cleanQueryResultsCache();
  }

  // Since postView calls postViewString, just run the test once
  @Test
  public void testPostView() throws Exception {
    String viewString = readXmlFromFile(VIEW_FILENAME_OK);
    View view = xmlToView(viewString);
    String location = ldmClientCentraxx.postView(view);
    assertFalse(LdmClientUtil.isNullOrEmpty(location));
  }

  @Test
  public void testGetResult() throws Exception {
    assumeFalse(LdmClientUtil.isNullOrEmpty(predefinedResultId));
    assertTrue(ldmClientCentraxx.getResult(predefinedResultLocation) instanceof QueryResult);
  }

  @Test
  public void testGetResultPageThreeTimes() throws Exception {
    assumeFalse(LdmClientUtil.isNullOrEmpty(predefinedResultId));
    Stopwatch stopwatch = Stopwatch.createStarted();
    assertTrue(ldmClientCentraxx.getResultPage(predefinedResultLocation, 0) instanceof QueryResult);
    logger.info("Getting page 0 took " + stopwatch.stop().elapsed(TimeUnit.MILLISECONDS) + " ms");
    stopwatch.reset().start();
    assertTrue(ldmClientCentraxx.getResultPage(predefinedResultLocation, 1) instanceof QueryResult);
    logger.info("Getting page 1 took " + stopwatch.stop().elapsed(TimeUnit.MILLISECONDS) + " ms");
    stopwatch.reset().start();
    assertTrue(ldmClientCentraxx.getResultPage(predefinedResultLocation, 0) instanceof QueryResult);
    logger.info("Getting page 0 took " + stopwatch.stop().elapsed(TimeUnit.MILLISECONDS) + " ms");
  }

  @Test
  public void testGetResultPage() throws Exception {
    assumeFalse(LdmClientUtil.isNullOrEmpty(predefinedResultId));
    assertTrue(ldmClientCentraxx.getResultPage(predefinedResultLocation, 0) instanceof QueryResult);
  }

  @Test
  public void testGetStatsOrError() throws Exception {
    assumeFalse(LdmClientUtil.isNullOrEmpty(predefinedResultId) && LdmClientUtil
        .isNullOrEmpty(predefinedErrorResultId));
    LdmQueryResult ldmQueryResult = ldmClientCentraxx.getStatsOrError(predefinedResultLocation);
    assertTrue(ldmQueryResult.hasResult() || ldmQueryResult.hasError());
  }

  @Test
  public void testGetQueryResultStatistic() throws Exception {
    assumeFalse(LdmClientUtil.isNullOrEmpty(predefinedResultId));
    assertNotNull(ldmClientCentraxx.getQueryResultStatistic(predefinedResultLocation));
  }

  @Test
  public void testGetError() throws Exception {
    assumeFalse(LdmClientUtil.isNullOrEmpty(predefinedErrorResultId));
    assertNotNull(ldmClientCentraxx.getError(predefinedErrorResultLocation));
  }

  @Test
  public void testGetVersionString() throws Exception {
    assertFalse(LdmClientUtil.isNullOrEmpty(ldmClientCentraxx.getVersionString()));
  }

  private String readXmlFromFile(String filename) {
    String xml = "";

    try (InputStream is = getClass().getClassLoader().getResourceAsStream(filename)) {
      xml = IOUtils.toString(is, StandardCharsets.UTF_8);
    } catch (IOException ioEx) {
      ioEx.printStackTrace();
    }

    return xml;
  }

  private View xmlToView(String xml) throws JAXBException {
    InputSource inputSource = new InputSource(new StringReader(xml));

    JAXBContext jaxbContext = JAXBContext.newInstance(View.class);
    Unmarshaller jaxbUnmarshaller = jaxbContext.createUnmarshaller();
    return (View) jaxbUnmarshaller.unmarshal(inputSource);
  }
  */

}


src/main/java/de/samply/common/ldmclient/centraxx/model/CentraxxInfo.java
package de.samply.common.ldmclient.centraxx.model;

/**
 * Created by michael on 22.02.17.
 */
public class CentraxxInfo {

  private String centraxxVersion;
  private String currentSessionCount;
  private String status;

  /**
   * Get the Centraxx version.
   * @return the centraxxVersion
   */
  public String getCentraxxVersion() {
    return centraxxVersion;
  }

  /**
   * Set the Centraxx version.
   * @param centraxxVersion the centraxxVersion to set
   */
  public void setCentraxxVersion(String centraxxVersion) {
    this.centraxxVersion = centraxxVersion;
  }

  /**
   * Get the current session count.
   * @return the currentSessionCount
   */
  public String getCurrentSessionCount() {
    return currentSessionCount;
  }

  /**
   * Set the current session count.
   * @param currentSessionCount the currentSessionCount to set
   */
  public void setCurrentSessionCount(String currentSessionCount) {
    this.currentSessionCount = currentSessionCount;
  }

  /**
   * Get the status.
   * @return the status
   */
  public String getStatus() {
    return status;
  }

  /**
   * Set the status.
   * @param status the status to set
   */
  public void setStatus(String status) {
    this.status = status;
  }

}


src/main/java/de/samply/common/ldmclient/centraxx/model/ObjectFactory.java
package de.samply.common.ldmclient.centraxx.model;

import javax.xml.bind.annotation.XmlRegistry;


/**
 * This object contains factory methods for each Java content interface and Java element interface
 * generated in the de.samply.share.model.upload.kairos package.
 * An ObjectFactory allows you to programatically.
 * construct new instances of the Java representation for XML content. The Java representation of
 * XML content can consist of schema derived interfaces and classes representing the binding of
 * schema type definitions, element declarations and model groups.  Factory methods for each of
 * these are provided in this class.
 */
@XmlRegistry
public class ObjectFactory {


  /**
   * Create a new ObjectFactory that can be used to create new instances of schema derived classes
   * for package: de.samply.share.model.upload.kairos.
   */
  public ObjectFactory() {
  }


  /**
   * Create an instance of {@link QueryResultStatistic }.
   * @return a empty QueryResultStatistic
   */
  public QueryResultStatistic createQueryResultStatistic() {
    return new QueryResultStatistic();
  }

}


src/main/java/de/samply/common/ldmclient/centraxx/utils/PercentageLogger.java
package de.samply.common.ldmclient.centraxx.utils;

import com.google.common.base.Preconditions;
import org.slf4j.Logger;

/**
 * PercentageLogger is intended to log the progress of another process, that consists of
 * "numberOfElements" steps and that is described through "description".
 *
 * <p>Example:</p>
 *
 * <p>doing something in several steps...</p>
 * <p>10%</p>
 * <p>40%</p>
 * <p>70%</p>
 * <p>90%</p>
 * <p>100%</p>
 */

public class PercentageLogger {

  private final Logger logger;
  private final int numberOfElements;
  private final String description;

  private int counter = 0;
  private int lastPercentage = 0;

  private boolean isStarted = false;

  /**
   * Calculates percentage of remaining steps and logs it.
   *
   * @param logger           logger
   * @param numberOfElements number of elements
   * @param description      description of the operation
   */
  public PercentageLogger(Logger logger, int numberOfElements, String description) {

    Preconditions.checkArgument(numberOfElements > 0, "numberOfElements is negative");

    this.logger = logger;
    this.numberOfElements = numberOfElements;
    this.description = description;

  }

  /**
   * Logs description.
   */
  public void start() {
    logger.debug(description);
    isStarted = true;
  }

  /**
   * Increments one step.
   */
  public void incrementCounter() {

    if (!isStarted) {
      start();
    }

    int percentage = 100 * ++counter / numberOfElements;

    if (lastPercentage != percentage) {

      lastPercentage = percentage;
      if (percentage % 10 == 0) {
        logger.debug(percentage + " %");
      }

    }

  }


}


src/main/java/de/samply/common/ldmclient/centraxx/LdmClientCentraxx.java
package de.samply.common.ldmclient.centraxx;

import com.google.gson.Gson;
import com.google.gson.JsonIOException;
import com.google.gson.JsonSyntaxException;
import de.samply.common.ldmclient.LdmClientException;
import de.samply.common.ldmclient.LdmClientUtil;
import de.samply.common.ldmclient.LdmClientView;
import de.samply.common.ldmclient.centraxx.model.CentraxxInfo;
import de.samply.common.ldmclient.model.LdmQueryResult;
import de.samply.share.common.utils.PercentageLogger;
import de.samply.share.model.ccp.QueryResult;
import de.samply.share.model.common.Error;
import de.samply.share.model.common.QueryResultStatistic;
import de.samply.share.model.common.View;
import de.samply.share.utils.QueryConverter;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.concurrent.TimeUnit;
import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Unmarshaller;
import org.apache.http.Consts;
import org.apache.http.HttpEntity;
import org.apache.http.HttpStatus;
import org.apache.http.StatusLine;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.HttpResponseException;
import org.apache.http.client.ResponseHandler;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.util.EntityUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 * Client to communicate with the local datamanagement implementation "Centraxx".
 */
public class LdmClientCentraxx extends
    LdmClientView<QueryResult, QueryResultStatistic, de.samply.share.model.ccp.Error,
        de.samply.share.model.ccp.View> {

  private static final String REST_PATH_TEILER = "teiler/";
  private static final String REST_PATH_INFO = "info/";
  private static final boolean CACHING_DEFAULT_VALUE = false;
  private final Logger logger = LoggerFactory.getLogger(LdmClientCentraxx.class);

  public LdmClientCentraxx(CloseableHttpClient httpClient, String centraxxBaseUrl)
      throws LdmClientException {
    this(httpClient, centraxxBaseUrl, CACHING_DEFAULT_VALUE);
  }

  public LdmClientCentraxx(CloseableHttpClient httpClient, String centraxxBaseUrl,
      boolean useCaching) throws LdmClientException {
    this(httpClient, centraxxBaseUrl, useCaching, CACHE_DEFAULT_SIZE);
  }

  public LdmClientCentraxx(CloseableHttpClient httpClient, String centraxxBaseUrl,
      boolean useCaching, int cacheSize) throws LdmClientException {
    super(httpClient, centraxxBaseUrl, useCaching, cacheSize);
  }

  public boolean isLdmCentraxx() {
    return true;
  }

  @Override
  protected String getFullPath(boolean statisticsOnly) {
    String path = REST_PATH_REQUESTS;
    if (statisticsOnly) {
      path = path + REST_RESULTS_ONLY_SUFFIX;
    }

    //return LdmClientUtil.addTrailingSlash(getLdmBaseUrl()) + REST_PATH_TEILER + path;
    return LdmClientUtil.addTrailingSlash(getLdmBaseUrl()) + path;
  }

  @Override
  protected Class<de.samply.share.model.ccp.View> getSpecificViewClass() {
    return de.samply.share.model.ccp.View.class;
  }

  @Override
  protected Class<QueryResult> getResultClass() {
    return QueryResult.class;
  }

  @Override
  protected Class<QueryResultStatistic> getStatisticsClass() {
    return QueryResultStatistic.class;
  }

  private Class getStatisticsClass(String response) {
    return (response.contains("http://de.kairos.centraxx/ccp/QueryResultStatistic"))
        ? de.samply.common.ldmclient.centraxx.model.QueryResultStatistic.class
        : getStatisticsClass();
  }

  @Override
  protected Class<de.samply.share.model.ccp.Error> getErrorClass() {
    return de.samply.share.model.ccp.Error.class;
  }

  @Override
  protected Class<?> getObjectFactoryClassForPostView() {
    return de.samply.share.model.ccp.ObjectFactory.class;
  }

  @Override
  protected Class<?> getObjectFactoryClassForResult() {
    return de.samply.share.model.ccp.ObjectFactory.class;
  }

  @Override
  protected de.samply.share.model.ccp.View convertCommonViewToSpecificView(View view)
      throws JAXBException {
    return QueryConverter.convertCommonViewToCcpView(view);
  }

  @Override
  protected View convertSpecificViewToCommonView(de.samply.share.model.ccp.View ccpView)
      throws JAXBException {
    return QueryConverter.convertCcpViewToCommonView(ccpView);
  }

  /**
   * Transforms a CCP QueryResultStatistic to a common QueryResultStatistic.
   *
   * @param queryResultStatistic the QueryResultStatistic in the centraxx namespace
   * @return the QueryResultStatistic in the common namespace
   */
  public LdmQueryResult convertQueryResultStatisticToCommonQueryResultStatistic(
      de.samply.common.ldmclient.centraxx.model.QueryResultStatistic queryResultStatistic) {
    QueryResultStatistic commonQueryResultStatistic = new QueryResultStatistic();
    commonQueryResultStatistic.setNumberOfPages(queryResultStatistic.getNumberOfPages());
    commonQueryResultStatistic.setRequestId(queryResultStatistic.getRequestId());
    commonQueryResultStatistic.setTotalSize(queryResultStatistic.getTotalSize());

    return new LdmQueryResult(commonQueryResultStatistic);
  }

  @Override
  protected LdmQueryResult convertQueryResultStatisticToCommonQueryResultStatistic(
      QueryResultStatistic qrs) throws JAXBException {
    return new LdmQueryResult(qrs);
  }


  @Override
  protected LdmQueryResult convertSpecificErrorToCommonError(de.samply.share.model.ccp.Error error)
      throws JAXBException {
    Error convertedError = QueryConverter.convertCcpErrorToCommonError(error);

    return new LdmQueryResult(convertedError);
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public QueryResult getResult(String location) throws LdmClientException {
    QueryResult queryResult = new QueryResult();
    QueryResultStatistic queryResultStatistic = getQueryResultStatistic(location);

    if (queryResultStatistic != null && queryResultStatistic.getTotalSize() > 0) {

      int numberOfPages = queryResultStatistic.getNumberOfPages();
      PercentageLogger percentageLogger = new PercentageLogger(logger, numberOfPages,
          "getting results from centraxx...");
      percentageLogger.start();

      for (int i = 0; i < numberOfPages; i++) {
        QueryResult queryResultPage = getResultPage(location, i);
        queryResult.getPatient().addAll(queryResultPage.getPatient());
        percentageLogger.incrementCounter();
      }
      queryResult.setId(queryResultStatistic.getRequestId());


    }
    return queryResult;
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public String getVersionString() throws LdmClientException {
    CentraxxInfo centraxxInfo = getInfo();
    return (centraxxInfo != null) ? centraxxInfo.getCentraxxVersion() : null;
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public String getUserAgentInfo() throws LdmClientException {

    String centraxxVersion = getVersionString();
    return (centraxxVersion != null) ? "Centraxx/" + centraxxVersion : "No CentraXX info provided";

  }

  // This class is intended for the compatibility with previous versions of centraXX 3.14
  // (and with the new ones)
  // This class is very similar to getStatsOrError defined in the library ldmclient
  // TODO: Delete this method after all DKTK-Sites use CentraXX 3.14
  @Override
  public LdmQueryResult getStatsOrError(String location) throws LdmClientException {
    HttpGet httpGet = new HttpGet(LdmClientUtil.addTrailingSlash(location) + REST_PATH_STATS);
    addHttpHeaders(httpGet);

    if (isLdmCentraxx()) {
      // Apparently, it may take a bit longer to reply when a new user session has to be created...
      // so use an extensive timeout of 1 minute
      RequestConfig requestConfig = RequestConfig.custom().setSocketTimeout(60000)
          .setConnectTimeout(60000).setConnectionRequestTimeout(60000).build();
      httpGet.setConfig(requestConfig);
    }

    try {
      CloseableHttpResponse response = getHttpClient().execute(httpGet);
      int statusCode = response.getStatusLine().getStatusCode();
      HttpEntity entity = response.getEntity();
      String entityOutput = EntityUtils.toString(entity, Consts.UTF_8);
      if (statusCode == HttpStatus.SC_OK) {

        Class statisticsClass = getStatisticsClass(entityOutput);

        JAXBContext jaxbContext = JAXBContext.newInstance(statisticsClass);
        Unmarshaller jaxbUnmarshaller = jaxbContext.createUnmarshaller();
        Object qrs = jaxbUnmarshaller.unmarshal(new StringReader(entityOutput));
        EntityUtils.consume(entity);
        response.close();

        return (qrs instanceof de.samply.common.ldmclient.centraxx.model.QueryResultStatistic)
            ? convertQueryResultStatisticToCommonQueryResultStatistic(
            (de.samply.common.ldmclient.centraxx.model.QueryResultStatistic) qrs) :
            convertQueryResultStatisticToCommonQueryResultStatistic((QueryResultStatistic) qrs);

      } else if (statusCode == HttpStatus.SC_UNPROCESSABLE_ENTITY) {
        JAXBContext jaxbContext = JAXBContext.newInstance(getErrorClass());
        Unmarshaller jaxbUnmarshaller = jaxbContext.createUnmarshaller();
        de.samply.share.model.ccp.Error error = (de.samply.share.model.ccp.Error) jaxbUnmarshaller
            .unmarshal(new StringReader(entityOutput));
        EntityUtils.consume(entity);
        response.close();
        return convertSpecificErrorToCommonError(error);
      } else if (statusCode == HttpStatus.SC_ACCEPTED) {
        response.close();
        logger
            .debug("Statistics not written yet. LDM client is probably busy with another request.");
        return LdmQueryResult.EMPTY;
      } else {
        response.close();
        throw new LdmClientException("Unexpected response code: " + statusCode);
      }
    } catch (IOException | JAXBException e) {
      throw new LdmClientException("While trying to read stats/error", e);
    }
  }

  private CentraxxInfo getInfo() throws LdmClientException {
    ResponseHandler<CentraxxInfo> responseHandler = response -> {
      StatusLine statusLine = response.getStatusLine();
      HttpEntity entity = response.getEntity();
      if (statusLine.getStatusCode() >= 300) {
        throw new HttpResponseException(
            statusLine.getStatusCode(),
            statusLine.getReasonPhrase());
      }
      if (entity == null) {
        throw new ClientProtocolException("Response contains no content");
      }
      try (InputStream instream = entity.getContent(); Reader reader = new InputStreamReader(
          instream, Consts.UTF_8)) {
        return new Gson().fromJson(reader, CentraxxInfo.class);
      } catch (JsonSyntaxException | JsonIOException e) {
        logger.warn("JSON Exception caught while trying to unmarshal centraxx info...");
        return null;
      }
    };

    String infoUrl = getInfoUrl();
    try {
      return getHttpClient().execute(new HttpGet(infoUrl), responseHandler);
    } catch (IOException e) {
      throw new LdmClientException("Could not read centraxx info", e);
    }
  }

  private String getInfoUrl() {

    //TODO : FIX URL
    String baseUrl = LdmClientUtil.addTrailingSlash(getLdmBaseUrl());
    if (baseUrl.contains(REST_PATH_TEILER)) {
      int index = baseUrl.indexOf(REST_PATH_TEILER);
      baseUrl = baseUrl.substring(0, index);
    }
    return baseUrl + REST_PATH_INFO;

  }

  /**
   * Todo.
   *
   * @param queryResult Todo.
   * @return Todo.
   * @throws JAXBException        Todo.
   * @throws LdmClientException   Todo.
   * @throws InterruptedException Todo.
   * @throws IOException          Todo.
   */
  public QueryResult exportQuery(QueryResult queryResult)
      throws JAXBException, LdmClientException, InterruptedException, IOException {
    ArrayList<String> patientids = new ArrayList<>();
    for (int i = 0; i < queryResult.getPatient().size(); i++) {
      patientids.add(queryResult.getPatient().get(i).getId());
    }
    StringBuilder builder = new StringBuilder();
    builder.append(
        "<ns5:View xmlns:ns2=\"http://schema.samply.de/common/MdrKey\" "
            + "xmlns:ns3=\"http://schema.samply.de/common/Value\" "
            + "xmlns:ns4=\"http://schema.samply.de/common/Attribute\" "
            + "xmlns:ns5=\"http://schema.samply.de/common/Query\">\n"
            + "<ns5:Query>\n"
            + "<ns5:Where>\n"
            + "<ns5:Or>\n");
    for (String patientid : patientids) {
      builder.append("<ns5:Eq>\n"
          + "<ns4:Attribute>\n"
          + "<ns2:MdrKey>urn:dktk:dataelement:91:1</ns2:MdrKey>\n"
          + "<ns3:Value>" + patientid + "</ns3:Value>\n"
          + "</ns4:Attribute>\n"
          + "</ns5:Eq>\n");
    }
    builder.append("</ns5:Or>\n"
        + "</ns5:Where>\n"
        + "</ns5:Query>\n"
        + "<ns5:ViewFields>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:54:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:26:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:1:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:24:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:25:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:46:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:43:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:45:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:74:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:72:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:73:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:80:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:99:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:79:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:101:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:78:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:100:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:2:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:10:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:81:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:18:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:82:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:28:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:21:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:29:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:77:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:9:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:3:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:8:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:5:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:19:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:4:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:98:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:7:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:6:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:83:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:89:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:39:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:38:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:69:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:23:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:67:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:40:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:36:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:33:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:34:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:48:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:53:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:49:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:90:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:50:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:97:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:dktk:dataelement:95:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:adt:dataelement:77:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:adt:dataelement:78:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:adt:dataelement:90:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:adt:dataelement:93:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:adt:dataelement:89:*</ns2:MdrKey>\n"
        + "<ns2:MdrKey>urn:adt:dataelement:91:*</ns2:MdrKey>\n"
        + "</ns5:ViewFields>\n"
        + "</ns5:View>");

    View v = QueryConverter.xmlToView(builder.toString());
    String location = this.postView(v);
    HttpGet httpGet = new HttpGet(location + "/" + REST_PATH_RESULT + REST_PARAM_PAGE + 0);
    RequestConfig requestConfig = RequestConfig.custom().setSocketTimeout(60000)
        .setConnectTimeout(60000).setConnectionRequestTimeout(60000).build();
    httpGet.setConfig(requestConfig);
    int statusCode = 202;
    while (statusCode == 202) {
      CloseableHttpResponse response = getHttpClient().execute(httpGet);
      statusCode = response.getStatusLine().getStatusCode();
      if (statusCode == 202) {
        TimeUnit.SECONDS.sleep(3);
      }
    }
    queryResult = this.getResult(location);
    System.out.println(builder.toString());
    return queryResult;
  }
}
